src/apps/MASS3DPA.hpp

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+///
+/// Action of 3D Mass matrix via partial assembly
+///
+/// Based on MFEM's/CEED algorithms.
+/// Reference implementation
+/// https://github.com/mfem/mfem/blob/master/fem/bilininteg_mass_pa.cpp#L925
+///
+///
+
+#ifndef RAJAPerf_Apps_MASS3DPA_HPP
+#define RAJAPerf_Apps_MASS3DPA_HPP
+
+#define MASS3DPA_DATA_SETUP \
+Real_ptr B = m_B; \
+Real_ptr Bt = m_Bt; \
+Real_ptr D = m_D; \
+Real_ptr X = m_X; \
+Real_ptr Y = m_Y; \
+Index_type NE = m_NE;
+
+#include "common/KernelBase.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+//Number of Dofs/Qpts in 1D
+#define D1D 4
+#define Q1D 5
+#define B_(x, y) B[x + Q1D * y]
+#define Bt_(x, y) Bt[x + D1D * y]
+#define X_(dx, dy, dz, e)                                                      \
+  X[dx + D1D * dy + D1D * D1D * dz + D1D * D1D * D1D * e]
+#define Y_(dx, dy, dz, e)                                                      \
+  Y[dx + D1D * dy + D1D * D1D * dz + D1D * D1D * D1D * e]
+#define D_(qx, qy, qz, e)                                                      \
+  D[qx + Q1D * qy + Q1D * Q1D * qz + Q1D * Q1D * Q1D * e]
+
+#define MASS3DPA_0_CPU           \
+        constexpr int MQ1 = Q1D; \
+        constexpr int MD1 = D1D; \
+        constexpr int MDQ = (MQ1 > MD1) ? MQ1 : MD1; \
+        double sDQ[MQ1 * MD1]; \
+        double(*Bsmem)[MD1] = (double(*)[MD1])sDQ; \
+        double(*Btsmem)[MQ1] = (double(*)[MQ1])sDQ; \
+        double sm0[MDQ * MDQ * MDQ]; \
+        double sm1[MDQ * MDQ * MDQ]; \
+        double(*Xsmem)[MD1][MD1] = (double(*)[MD1][MD1])sm0; \
+        double(*DDQ)[MD1][MQ1] = (double(*)[MD1][MQ1])sm1; \
+        double(*DQQ)[MQ1][MQ1] = (double(*)[MQ1][MQ1])sm0; \
+        double(*QQQ)[MQ1][MQ1] = (double(*)[MQ1][MQ1])sm1; \
+        double(*QQD)[MQ1][MD1] = (double(*)[MQ1][MD1])sm0; \
+        double(*QDD)[MD1][MD1] = (double(*)[MD1][MD1])sm1;
+
+#define MASS3DPA_0_GPU \
+        constexpr int MQ1 = Q1D; \
+        constexpr int MD1 = D1D; \
+        constexpr int MDQ = (MQ1 > MD1) ? MQ1 : MD1; \
+        RAJA_TEAM_SHARED  double sDQ[MQ1 * MD1];     \
+        double(*Bsmem)[MD1] = (double(*)[MD1])sDQ; \
+        double(*Btsmem)[MQ1] = (double(*)[MQ1])sDQ; \
+        RAJA_TEAM_SHARED double sm0[MDQ * MDQ * MDQ];       \
+        RAJA_TEAM_SHARED double sm1[MDQ * MDQ * MDQ];      \
+        double(*Xsmem)[MD1][MD1] = (double(*)[MD1][MD1])sm0; \
+        double(*DDQ)[MD1][MQ1] = (double(*)[MD1][MQ1])sm1; \
+        double(*DQQ)[MQ1][MQ1] = (double(*)[MQ1][MQ1])sm0; \
+        double(*QQQ)[MQ1][MQ1] = (double(*)[MQ1][MQ1])sm1; \
+        double(*QQD)[MQ1][MD1] = (double(*)[MQ1][MD1])sm0; \
+        double(*QDD)[MD1][MD1] = (double(*)[MD1][MD1])sm1;
+
+#define MASS3DPA_1 \
+  RAJA_UNROLL(MD1) \
```

Comment:
where is this "unroll" macro defined?

Reply:
Its defined in the back-end specific variants. Depending on which back-end we use; unrolling loops can help or hurt performance; example CUDA/HIP so we either do or do not unroll.